File: 4.8-bc/go/stage-3-hexagonal/billing/infra/http/handler.go
package http

import (
	"errors"
	stdhttp "net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/billing/app"
	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/billing/domain"
	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/shared/apperr"
	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/shared/httputil"
)

type Handler struct {
	svc *app.Service
}

func NewHandler(svc *app.Service) *Handler {
	return &Handler{svc: svc}
}

func (h *Handler) Register(r chi.Router) {
	r.Post("/subscriptions", h.subscribe)
}

type subscribeReq struct {
	UserID string `json:"user_id"`
	Plan   string `json:"plan"`
}

type subscribeResp struct {
	SubscriptionID uuid.UUID `json:"subscription_id"`
	Plan           string    `json:"plan"`
	NextChargeAt   time.Time `json:"next_charge_at"`
}

func (h *Handler) subscribe(w stdhttp.ResponseWriter, r *stdhttp.Request) {
	var req subscribeReq
	if err := httputil.DecodeBody(r, &req); err != nil {
		httputil.WriteJSON(w, stdhttp.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	uid, err := uuid.Parse(req.UserID)
	if err != nil {
		httputil.WriteJSON(w, stdhttp.StatusBadRequest, map[string]string{"error": "invalid_user_id"})
		return
	}
	sub, err := h.svc.Subscribe(r.Context(), uid, req.Plan)
	if err != nil {
		if errors.Is(err, domain.ErrInvalidPlan) {
			httputil.WriteError(w, apperr.New("billing.invalid_plan", "plan must be basic|pro|enterprise", stdhttp.StatusBadRequest))
			return
		}
		httputil.WriteError(w, err)
		return
	}
	httputil.WriteJSON(w, stdhttp.StatusCreated, subscribeResp{
		SubscriptionID: sub.ID,
		Plan:           sub.Plan,
		NextChargeAt:   sub.NextChargeAt,
	})
}

File: 4.8-bc/ts/stage-3-hexagonal/src/billing/infra/http/handler.ts
// HTTP adapter для Billing BC.
import type { FastifyInstance } from "fastify";

import { AppError } from "../../../shared/apperr.js";
import { writeError } from "../../../shared/httputil.js";
import type { Service } from "../../app/service.js";
import { InvalidPlanError } from "../../domain/errors.js";

interface SubscribeReq {
  user_id?: string;
  plan?: string;
}

interface SubscribeResp {
  subscription_id: string;
  plan: string;
  next_charge_at: Date;
}

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export class Handler {
  constructor(private readonly svc: Service) {}

  register(app: FastifyInstance): void {
    app.post<{ Body: SubscribeReq }>("/subscriptions", async (req, reply) => {
      const body = req.body ?? {};
      if (!body.plan) {
        return reply.code(400).send({ error: "invalid_request" });
      }
      if (!body.user_id || !UUID_RE.test(body.user_id)) {
        return reply.code(400).send({ error: "invalid_user_id" });
      }
      try {
        const sub = await this.svc.subscribe(body.user_id, body.plan);
        const resp: SubscribeResp = {
          subscription_id: sub.id,
          plan: sub.plan,
          next_charge_at: sub.nextChargeAt,
        };
        return reply.code(201).send(resp);
      } catch (err) {
        if (err instanceof InvalidPlanError) {
          return writeError(
            reply,
            new AppError(
              InvalidPlanError.CODE,
              "plan must be basic|pro|enterprise",
              400,
            ),
          );
        }
        req.log.error({ err }, "subscribe failed");
        return writeError(reply, err);
      }
    });
  }
}
